flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/Regularization.scala

```diff
@@ -0,0 +1,207 @@
+    */
+  def regGradient(
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regParameter: Double)
+}
+
+/** Performs no regularization, equivalent to $R(w) = 0$ **/
+class NoRegularization extends Regularization {
+  /** Adds the regularization term to the loss value
+    *
+    * @param loss The loss value, before applying regularization
+    * @param weightVector The current vector of weights
+    * @param regParameter The regularization parameter, $\lambda$
+    * @return The loss value with regularization applied.
+    */
+  override def regLoss(
+    loss: Double,
+    weightVector: FlinkVector,
+    regParameter: Double):  Double = {loss}
+}
+
+/** $L_2$ regularization penalty.
+  *
+  * Penalizes large weights, favoring solutions with more small weights rather than few large ones.
+  *
+  */
+class L2Regularization extends DiffRegularization {
+
+  /** Adds the regularization term to the loss value
+    *
+    * @param loss The loss value, before applying regularization
+    * @param weightVector The current vector of weights
+    * @param regParameter The regularization parameter, $\lambda$
+    * @return The loss value with regularization applied.
+    */
+  override def regLoss(loss: Double, weightVector: FlinkVector, regParameter: Double)
+    : Double = {
+    val brzVector = weightVector.asBreeze
+    loss + regParameter * (brzVector dot brzVector) / 2
+  }
+
+  /** Adds the regularization gradient term to the loss gradient. The gradient is updated in place.
+    *
+    * @param weightVector The current vector of weights.
+    * @param lossGradient The loss gradient, without regularization. Updated in-place.
+    * @param regParameter The regularization parameter, $\lambda$.
+    */
+  override def regGradient(
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regParameter: Double): Unit = {
+    BLAS.axpy(regParameter, weightVector, lossGradient)
+  }
+}
+
+/** $L_1$ regularization penalty.
+  *
+  * The $L_1$ penalty can be used to drive a number of the solution coefficients to 0, thereby
+  * producing sparse solutions.
+  *
+  */
+class L1Regularization extends Regularization {
+  /** Calculates and applies the regularization amount and the regularization parameter
+    *
+    * Implementation was taken from the Apache Spark Mllib library:
+    * http://git.io/vfZIT
+    *
+    * @param oldWeights The weights to be updated
+    * @param gradient The gradient according to which we will update the weights
+    * @param effectiveStepSize The effective step size for this iteration
+    * @param regParameter The regularization parameter to be applied in the case of L1
+    *                     regularization
+    */
+  override def takeStep(
+      oldWeights: FlinkVector,
+      gradient: FlinkVector,
+      effectiveStepSize: Double,
+      regParameter: Double) {
+    BLAS.axpy(-effectiveStepSize, gradient, oldWeights)
+    val brzWeights = oldWeights.asBreeze
```

Comment:
Why do we need to operate on Breeze vectors? Element-wise modification should also be possible with Flink vectors. That way, we don't have a Breeze dependency in our optimization framework.